### ISSUE TITLE
Provide mixin alternative to add adapter functionality (or do complete switch to mixin)

### ISSUE THREAD

**Issue by kennethdavidbuck** [thaume/ember-cli-uuid#15]
First off, thank you for this addon, I really appreciate the work that you have put into it!

Right now the addon is dependent on an app initializer to reopen the ember data adapter and override the generateIdForRecord hook. This is problematic since it means unit tests etc. end up missing the feature. Further to this, the initializer is under app, so it can't be imported and run. It would be great if the functionality was available as a mixin that consumers could add to their application, or more specific, adapter instead. Doing it this way would only require a little more prep work, but would offer more control and make adapter behaviour consistent across each test type.

I would be more than happy to do a PR for this if you are interested.

thank you for your consideration!

**Comment by thaume** [comment]
Hello @kennethdavidbuck and thanks for the kind words !

This is interesting since it is the way things were working in the very beginning (adding a mixin to your adapter). If I remember correctly I switched to an "auto uuid" to make the addon easier to use :)

How would you see that working ? Do we want to block the initializer behaviour if the adapter already has the right method implemented (via the mixin) ? Through duck typing ?

I guess the only thing is that I would not want to break the way it currently works so that this update makes sense for people that need to use it through a mixin, but does not change anything for people using the "auto uuid" feature right now.

If you can work on a PR, that would be fabulous :) the code is pretty simple, but I can always help if you need !

Cheers !

**Comment by kennethdavidbuck** [comment]
Thank you for the quick reply! I was thinking about introducing an EXTEND_PROTOTYPES container, like ember has, under the 'ember-cli-uuid' namespace in the environment config. I would have the value for DS.Adapter be defaulted to true.
ex. 
```javascript
'ember-cli-uuid': {
    EXTEND_PROTOTYPES: {
        'DS.ADAPTER': true
   }
}
```
I could then create a mixin for the adapter that the app initializer could use when reopening the class. This would provide the ability for consumers to opt out of the auto-extension and without creating a breaking change. How does that sound?

**Comment by thaume** [comment]
That sounds like a good plan !

Let me know how I can help along the way :)

**Comment by kennethdavidbuck** [comment]
Cool. I do have one question. I see that you check whether defaultUUID is set to true inside of the #generateForId adapter rather than having the check wrap the reopening of the class. Is it an intended feature to allow the configuration to be updated at run time? If not, then I was thinking that we could reuse the existing defaultUUID configuration value rather than introducing the EXTEND_PROTOTYPES container, but have the check wrap the reopening of the class instead. Thoughts?

**Comment by thaume** [comment]
I think that the `defaultUUID` serves a different use case. We built it (#8 ) to let people use the UUID features through manual import + manual generation (when they only want to use it for some models, or even for something that is not a model).

EXTEND_PROTOTYPE would be a different use case where the initializer would not be executed (or at least the `DS.Adapter` would not be reopened).

This way we would have :

**Comment by kennethdavidbuck** [comment]
I think that sounds great. I'll should have something in the next few days here. Cheers!

**Comment by thaume** [comment]
Great to hear :)

**Comment by thaume** [comment]
Fixed by #16
